### PR TITLE
go/storage/mkvs: Use cbor.UnmarshalTrusted for internal metadata

### PR DESCRIPTION
**Pull request by kostko**

### CODE REVIEW

File: go/common/cbor/cbor.go [hunk 1]
```diff
@@ -27,14 +27,23 @@ var (
...
 		IndefLength: cbor.IndefLengthForbidden,
 		TagsMd:      cbor.TagsForbidden,
 	}
 
-	encMode cbor.EncMode
-	decMode cbor.DecMode
+	// decOptionsTrusted are decoding options for TRUSTED inputs. They are only used when explicitly
+	// requested by using the UnmarshalTrusted method.
+	decOptionsTrusted = cbor.DecOptions{
+		MaxArrayElements: 2097152,
```
Comment (ptrus) [inline go/common/cbor/cbor.go:40]: Maybe add a comment on why this number? Any reason to not just use the max allowed?

**Comment by kostko** [comment]
> hm, i guess it's already tested in the upstream lib, maybe a test case for the storage code that would fail with old code.

Yeah we don't currently have CBOR tests as those are well tested upstream. We could add some basic API tests for CBOR. But for the database case the problem is that we would need to generate huge data sets (e.g., trees with more than 132k updated nodes) to trigger this which would probably slow tests down. I'll try and see how that would look like.

File: go/common/cbor/cbor.go (hunk 1 above)
Comment (kostko) [reply go/common/cbor/cbor.go:40]: No special reason, I just multiplied the previous number by 16, could use the max yeah.

**Comment by kostko** [comment]
Ok, it's not that bad, 132k elements takes 2 seconds to process, commit and finalize (locally, probably more on CI?) so we can leave this test in.

File: go/common/cbor/cbor.go [hunk 2]
```diff
@@ -27,14 +27,23 @@ var (
...
 		TagsMd:      cbor.TagsForbidden,
 	}
 
-	encMode cbor.EncMode
-	decMode cbor.DecMode
+	// decOptionsTrusted are decoding options for TRUSTED inputs. They are only used when explicitly
+	// requested by using the UnmarshalTrusted method.
+	decOptionsTrusted = cbor.DecOptions{
+		MaxArrayElements: 134217728, // Maximum allowed.
+		MaxMapPairs:      134217728, // Maximum allowed.
```
Comment (pro-wh) [inline go/common/cbor/cbor.go:41]: what are the defaults for these on the untrusted `decOptions`?

Comment (kostko) [reply go/common/cbor/cbor.go:41]: [The default is 131072.](https://github.com/fxamacker/cbor#decoding-options) Maybe we should consider making this explicit and/or even changing it (e.g., lowering it) for untrusted settings?

Comment (pro-wh) [reply go/common/cbor/cbor.go:41]: eh, maybe a comment if anything

Comment (ridicoulous) [reply go/common/cbor/cbor.go:41]: @kostko hello. I have an issue with calling [StateToGenesis beacon api](https://github.com/oasisprotocol/oasis-core/blob/9b98fdcad2c64935df98a98a9aba55de9cbe0f06/go/beacon/api/grpc.go#L219): `rpc error: code = Internal desc = grpc: failed to unmarshal the received message cbor: exceeded max number of key-value pairs 131072 for CBOR map` at block `7188323`. How I can explicitly call UnmarshalTrusted method at this point?

**Pull request merged**